fido2/ctap.c

```diff
@@ -1154,6 +1157,270 @@ uint8_t ctap_get_next_assertion(CborEncoder * encoder)
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, "name");
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, rk.user.name);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, "displayName");
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&usr, rk.user.displayName);
+        check_ret(ret);
+        ret = cbor_encoder_close_container(&map, &usr);
+        check_ret(ret);
+    }
+
+    ret = cbor_encode_int(&map, 7);
+    check_ret(ret);
+    {
+        CborEncoder credId;
+        ret = cbor_encoder_create_map(&map, &credId, 1);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&credId, "id");
+        check_ret(ret);
+        ret = cbor_encode_byte_string(&credId, (uint8_t*)&rk.id, sizeof(CredentialId));
+        check_ret(ret);
+        ret = cbor_encoder_close_container(&map, &credId);
+        check_ret(ret);
+    }
+
+    ret = cbor_encode_int(&map, 8);
+    check_ret(ret);
+    ctap_generate_cose_key(&map, (uint8_t*)&rk.id, sizeof(CredentialId), PUB_KEY_CRED_PUB_KEY, COSE_ALG_ES256);
+
+    if (rk_count > 0)
+    {
+        ret = cbor_encode_int(&map, 9);
+        check_ret(ret);
+        ret = cbor_encode_int(&map, rk_count);
+        check_ret(ret);
+    }
+    ret = cbor_encoder_close_container(encoder, &map);
+    check_ret(ret);
+    return 0;
+}
+
+uint8_t ctap_cred_mgmt(CborEncoder * encoder, uint8_t * request, int length)
+{
+    CTAP_credMgmt CM;
+    CTAP_residentKey rk;
+    int i = 0;
+    // use the same index for both RP and RK commands, it make things simpler
+    static int curr_rk_ind = 0;
+    // keep the rpIdHash specified in CM_cmdRKBegin cause it's not present in CM_cmdRKNext
+    static uint8_t rpIdHash[32];
+    // number of stored RPs
+    int rp_count = 0;
+    // number of RKs with the specified rpIdHash
+    int rk_count = 0;
+
+    int ret = ctap_parse_cred_mgmt(&CM, request, length);
+    if (ret != 0)
+    {
+        printf2(TAG_ERR,"error, ctap_parse_cred_mgmt failed\n");
+        return ret;
+    }
+    if (CM.cmd == CM_cmdMetadata || CM.cmd == CM_cmdRPBegin)
+    {
+        if (CM.pinProtocol != 1)
+        {
+            return CTAP1_ERR_OTHER;
+        }
+        uint8_t cmd = (uint8_t) CM.cmd;
+        ret = verify_pin_auth_ex(CM.pinAuth, &cmd, 1);
+    }
+    if (CM.cmd == CM_cmdRKBegin)
+    {
+        uint8_t params[5 + sizeof(CM.rpIdHash)] = {CM.cmd, 0xa1, 0x01, 0x58, 0x20};
+        if (CM.pinProtocol != 1)
+        {
+            return CTAP1_ERR_OTHER;
+        }
+        memcpy(&params[5], CM.rpIdHash, sizeof(CM.rpIdHash));
+        ret = verify_pin_auth_ex(CM.pinAuth, params, sizeof(params));
+    }
+    if (ret == CTAP2_ERR_PIN_AUTH_INVALID)
+    {
+        ctap_decrement_pin_attempts();
+        if (ctap_device_boot_locked())
+        {
+            return CTAP2_ERR_PIN_AUTH_BLOCKED;
+        }
+        return CTAP2_ERR_PIN_AUTH_INVALID;
+    }
+    else
+    {
+        ctap_reset_pin_attempts();
```

Comment:
If an unsolicited `CM_cmdRPNext` or `CM_cmdRKNext` was sent, this would trigger the retries counter to get reset?  Then there would be a way for malware to have unlimited PIN attempts.  Ideally this should only done for commands that need pinAuth, and there should be a check that the "next" commands come after a valid command.

Reply:
Good catch. I will rework the pinAuth handling and add tests for these scenarios.

Reply:
I have extracted the pinAuth handling into a separate function and introduced some flags to guard against calling Next commands without corresponding Begin. Tests are also updated.

Reply:
I think this is still vulnerable, but now if an invalid sub command is sent, it will reset the pin attempts.  Maybe all of the attempt decrement + reset logic should be placed in the `verify_pin_auth_ex` function to be safe?

Reply:
`verify_pin_auth_ex` is called by `verify_pin_auth` so if I move the checks there it will have implications to the rest of the code.
I will change line 1299 to check if the command should have pinAuth and return if this is not the case.

Reply:
Ok, looks good!